hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala

```diff
@@ -367,4 +383,79 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
       destOff
     )
   }
+
+  override def storeShallowAtOffset(destOffset: Code[Long], valueAddress: Code[Long]): Code[Unit] = {
+    Region.storeAddress(destOffset, valueAddress)
+  }
+
+  // semantically this function expects a non-null sourceAddress, and by that property, this function results in a non-null value
+  override def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcAddress: Code[Long],
+  allowDowncast: Boolean = false, forceDeep: Boolean = false): Code[Long] = {
+    assert(srcPType.isInstanceOf[PArray])
+
+    val sourceType = srcPType.asInstanceOf[PArray]
+    val sourceElementType = sourceType.elementType.fundamentalType
```

Comment:
same question as above regarding checking equality on fundamentalTypes, note below my shallow copy check does use the fundamental version of src/dest element ptypes, not the element ptypes themselves.

It's also a cheaper check here, and 2 lines to change it back.